bundles/org.openhab.binding.magentatv/src/main/java/org/openhab/binding/magentatv/internal/handler/MagentaTVHandler.java

```diff
@@ -0,0 +1,692 @@
+            control.subscribeEventChannel();
+            control.sendPairingRequest();
+
+            // check for pairing timeout
+            final int iRefresh = ++idRefresh;
+            pairingWatchdogJob = scheduler.schedule(() -> {
+                if (iRefresh == idRefresh) { // Make a best effort to not run multiple deferred refresh
+                    if (config.getVerificationCode().isEmpty()) {
+                        setOnlineState(ThingStatus.OFFLINE, "Timeout on pairing request!");
+                    }
+                }
+            }, 15, TimeUnit.SECONDS);
+        }
+    }
+
+    /**
+     * If userID is empty and credentials are given the Telekom OAuth service is
+     * used to query the userID
+     *
+     * @throws MagentaTVException
+     */
+    private void authenticateUser() throws MagentaTVException {
+        String userId = config.getUserID();
+        if (userId.isEmpty()) {
+            // run OAuth authentication, this finally provides the userID
+            userId = control.authenticateUser(config.getAccountName(), config.getAccountPassword());
+
+            // Update thing configuration (persistent) - remove credentials, add userID
+            Configuration configuration = this.getConfig();
+            configuration.remove(PROPERTY_ACCT_NAME);
+            configuration.remove(PROPERTY_ACCT_PWD);
+            configuration.remove(PROPERTY_USERID);
+            configuration.put(PROPERTY_ACCT_NAME, "");
+            configuration.put(PROPERTY_ACCT_PWD, "");
+            configuration.put(PROPERTY_USERID, userId);
+            this.updateConfiguration(configuration);
+            config.setAccountName("");
+            config.setAccountPassword("");
+        } else {
+            logger.debug("{}: Skip OAuth, use existing userID {}", thingId, config.getUserID());
+        }
+        if (!userId.isEmpty()) {
+            config.setUserID(userId);
+        } else {
+            logger.warn("{}: Unable to obtain userId from OAuth", thingId);
+        }
+    }
+
+    /**
+     * Update thing status
+     *
+     * @param mode new thing status
+     * @return ON = power on, OFF=power off
+     */
+    public void setOnlineState(ThingStatus newStatus, String errorMessage) {
+        ThingStatus status = this.getThing().getStatus();
+        if (status != newStatus) {
+            if (newStatus == ThingStatus.ONLINE) {
+                updateStatus(newStatus);
+                updateState(CHANNEL_POWER, OnOffType.ON);
+            } else {
+                if (!errorMessage.isEmpty()) {
+                    logger.debug("{}: Communication Error - {}, switch Thing offline", thingId, errorMessage);
+                    updateStatus(newStatus, ThingStatusDetail.COMMUNICATION_ERROR, errorMessage);
+                } else {
+                    updateStatus(newStatus);
+                }
+                updateState(CHANNEL_POWER, OnOffType.OFF);
```

Comment:
This should not be set here - if communication breaks down, it does not automatically mean that the receiver has been switched off.

Reply:
The MR400 doesn't send the SSDN message, so in most cases the receiver is off when communication breaks down (usually the receiver is connected with ethernet). I spend a lot of time trying to get a better status handling and this was the outcome. Maybe not perfect, but works on a daily basis pretty well.

Reply:
Well, let's keep it that way as an unusual exception, but don't tell anyone ;-)